mesh/mesh_readers.cpp

```diff
@@ -917,6 +918,19 @@ void Mesh::ReadGmshMesh(std::istream &input, int &curved, int &read_gf)
    // (there may be gaps in the numbering, and also Gmsh enumerates vertices
    // starting from 1, not 0)
    map<int, int> vertices_map;
+
+   // Gmsh always outputs coordinates in 3D
+   spaceDim = 3;
```

Comment:
This causes 2D meshes to be read as surface meshes in MFEM, which is not what we want. Is there a way to check if the Gmsh file is actually a 2D mesh and set the `spaceDim` to 2 in that case?

Reply:
We could scan the elements looking for 3D elements but that would not guarantee 2D geometry.  We would also need to scan the z coordinates and verify that they lie on a plane (within some tolerance).

Reply:
How about we just check that the z-coordinate if the nodes is (identically) zero? It is strange that there is no way to distinguish between 2D and 3D in the Gmsh file :(

Reply:
I took a closer look at the `spaceDim` member data when reading Gmsh files.  Currently the master branch of MFEM reads in 3 coordinates and stores them in `Vertex` objects. However, the `spaceDim` is not set until `FinalizeTopology` is called where it is given a default value equal to `Dim`.  Technically this is a bug because the z coordinates could be non trivial but we are ignoring them.  This does, however, explain why 2D Gmsh meshes appear as 2D meshes rather than surface meshes.

Reply:
I can add code to compute the bounding box of the mesh as the nodes are read.  We can then set the space dimension based primarily on the element types and secondarily on the relative sizes of the bounding box.